staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/validator.go

```diff
@@ -101,30 +105,45 @@ func (v *validator) Validate(ctx context.Context, versionedAttr *admission.Versi
 			continue
 		}
 
+		var messageResult *cel.EvaluationResult
+		if len(messageResults) > i {
+			messageResult = &messageResults[i]
+		}
 		if evalResult.Error != nil {
 			decision.Action = policyDecisionActionForError(f)
 			decision.Evaluation = EvalError
 			decision.Message = evalResult.Error.Error()
+		} else if messageResult != nil {
+			err := messageResult.Error
+			if err != nil && strings.HasPrefix(err.Error(), "operation cancelled: actual cost limit exceeded") {
+				decision.Action = policyDecisionActionForError(f)
+				decision.Evaluation = EvalError
+				decision.Message = fmt.Sprintf("failed messageExpression: %s", err)
+			}
 		} else if evalResult.EvalResult != celtypes.True {
 			decision.Action = ActionDeny
 			if validation.Reason == nil {
 				decision.Reason = metav1.StatusReasonInvalid
 			} else {
 				decision.Reason = *validation.Reason
 			}
-			if len(validation.Message) > 0 {
-				decision.Message = strings.TrimSpace(validation.Message)
+			message := ""
+			if messageResult != nil && messageResult.Error == nil && messageResult.EvalResult != nil {
+				message = messageResult.EvalResult.Value().(string)
```

Comment:
check if messageExpression evaluated to "" or nil? (in this case should fall back to use `message` as well)

Reply:
Maybe check if it resolved to `string` type first?

Reply:
update the code to reflect the new behaviour.